src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -375,127 +330,206 @@ let ApplyAllOptimizations
-                implFile
-                |> InnerLambdasToTopLevelFuncs.MakeTopLevelRepresentationDecisions ccu tcGlobals
-            else
-                implFile
-
-        let implFile = LowerCalls.LowerImplFile tcGlobals implFile
-
-        if tcConfig.doFinalSimplify then
-            let (optEnvFinalSimplify, implFile, _, _), _ =
-                Optimizer.OptimizeImplFile(
-                    phase2And3Settings,
-                    ccu,
-                    tcGlobals,
-                    tcVal,
-                    importMap,
-                    env,
-                    isIncrementalFragment,
-                    tcConfig.fsiMultiAssemblyEmit,
-                    tcConfig.emitTailcalls,
-                    hidden,
-                    implFile
-                )
-
-            optEnvFinalSimplify, implFile
-        else
-            env, implFile
+    let measurements = ConcurrentDictionary<string, int64>()
 
-    let results, optEnvFirstLoop =
-        match tcConfig.optSettings.processingMode with
-        | Optimizer.OptimizationProcessingMode.PartiallyParallel ->
-            let ct = CancellationToken.None
+    let measure (name: string) =
+        let sw = Stopwatch.StartNew()
 
-            let results, optEnvFirstPhase =
-                ParallelOptimization.optimizeFilesInParallel optEnv (phase1, phase2, phase3) implFiles ct
+        { new System.IDisposable with
+            member this.Dispose() =
+                lock measurements (fun () ->
+                    if measurements.ContainsKey name = false then
+                        measurements[name] <- 0L
 
-            results |> Array.toList, optEnvFirstPhase
-        | Optimizer.OptimizationProcessingMode.Sequential ->
-            let results, (optEnvFirstLoop, _, _, _) =
-                ((optEnv, optEnv, optEnv, SignatureHidingInfo.Empty), implFiles)
+                    measurements[name] <- measurements[name] + sw.ElapsedMilliseconds)
+        }
 
-                ||> List.mapFold (fun (optEnvFirstLoop, optEnvExtraLoop, optEnvFinalSimplify, hidden) implFile ->
-                    let (optEnvFirstLoop, implFile, implFileOptData, hidden), optimizeDuringCodeGen =
-                        phase1 (optEnvFirstLoop, hidden, implFile)
+    let wrapPhaseFunc (f: PhaseFunc) (info: Phase) =
+        fun (inputs: PhaseInputs) ->
+            use _ = measure info.Name
 
-                    let optEnvExtraLoop, implFile =
-                        phase2 (optEnvExtraLoop, hidden, implFileOptData, implFile)
+            use _ =
+                let tags =
+                    [|
+                        "QualifiedNameOfFile", inputs.File.QualifiedNameOfFile.Text
+                        "OptimisationPhase", info.Name
+                    |]
 
-                    let optEnvFinalSimplify, implFile = phase3 (optEnvFinalSimplify, hidden, implFile)
+                FSharp.Compiler.Diagnostics.Activity.start $"file-{inputs.FileIdx}_phase-{info.Name}" tags
```

Comment:
```suggestion
                FSharp.Compiler.Diagnostics.Activity.start $"file-{inputs.FileIdx}_optimization_phase-{info.Name}" tags
```

Reply:
All spans are all nested in the parent "Optimizations" span, so the "optimization" bit is redundant and it makes the element longer in the UI - but I'm happy to change it if people insist.